beps/0002-dynamic-frontend-plugins/README.md

```diff
@@ -154,12 +154,386 @@ Each plugin is required to provide manifest file (metadata) in predefined format
+
+A dynamic plugin requires all dependencies to be present in the build output. Module sharing is then used to optimize the "output size" at runtime. If a shared package is not found in the browser scope, or if the version range does not match the requirement, it falls back and loads additional JS assets from the "private" build and adds them to the scope.
+
+Webpack is also not the best tool for building traditional NPM packages. Dependencies can be externalized, but Webpack brings additional code into the bundles. Using Webpack external dependencies also conflicts with module federation shared modules. Externals will not be included even as a shared module fallback. As a result, such shared module can't be used because the dependency will not be defined in the init container.
+
+If the vision is to still support "static" plugins installed via NPM, it makes sense to keep the static and dynamic plugin builds separate and build two different bundles. Both can still be published as a single NPM package (not recommended), but to achieve optimal performance for both, there cannot be a shared output.
+
+#### Dynamic export CLI
+
+```sh
+backstage-cli plugin export-dynamic
+```
+
+Similar to the current build command, instead of producing static plugin output, it builds a dynamic plugin using Webpack and module federation plugins.
+
+**Arguments**
+
+- _path_ (optional)
+  - path to the plugin `package.json`
+  - defaults to `cwd`
+- _export-name_
+  - Export name from where the plugin root is accessible
+  - defaults to `pluginEntry`
+- _cwd_
+  - current working directory of the script
+  - defaults to current `cwd` value
+
+**Shared modules**
+
+> NOTE: Details of module sharing are described in the [Module sharing](#module-sharing) section.
+
+Short summary of dependency sharing:
+
+- Predefined set of singleton packages
+- Generated granular config of modules for MUI dependencies
+- Share all other dependencies listed in package.json
+
+To allow additional customization of module sharing, extra configuration will be required. A similar approach to the current catalog-info.yaml config file can be used. There is also the option of embedding the metadata in the `package.json`. However, the configuration can become very large and clutter the package.json.
+
+- [Schema proposal draft](./dependency-sharing-schema.json)
+- [Config draft](./share-config-sample.yaml)
+
+#### Generate module sharing map
+
+> NOTE: RedHat has similar internal generators that are used in combination with Scalprum. These can be built upon, moved to Scalprum, and exposed. If there is interest, Spotify does not have to be responsible for maintaining the source code.
+
+A simple program to traverse the AST and find exports from the `main` or `module` script of a dependency. Any exposed modules from the dependency entry will be traced to their source files, and a module map can be generated.
```

Comment:
Is it important enough to find modules that are truly imported as opposed to simply including all transitive `package.json` dependencies for this?

Reply:
Most package.json dependencies do not matter and can be shared as a whole. There is no point looking for exports for `react` or `react-router-dom`. In fact, we do not share as a whole package, the module sharing breaks.

However specifically for MUI packages (and any other component libraries), it is critical to target specifically modules that are used. Sharing the whole shebang means the entire thing ends up in the bundle (can't tree share shared module). MUI icons package has `21226` modules (at the time of writing). That's 9.1 megabytes of JS. You have another 8 for the `@mui/material`. AFAIK, you also have the older version of MUI dependencies. When you load 4-10 (educated guess, way more for pinned dependencies) versions in a session where the sharing is not optimized you can have hundreds of megabytes of JS just for some icons and buttons. In that scenario, it is better not to share the dependency.

Reply:
Yes exactly. The AST I am proposing generates the module map. I see no harm in exposing the generator so it can be used for other dependencies as well. I don't think we want developers to construct MUI like config manually.